optimade/adapters/base.py

```diff
@@ -0,0 +1,75 @@
+from typing import Union, Dict, Callable, Any
+
+from optimade.models import EntryResource
+
+
+class EntryAdapter:
+    """Base class for lazy resource entry adapters
+    :param entry: JSON OPTIMADE single resource entry.
+    """
+
+    ENTRY_RESOURCE: EntryResource = EntryResource
+    _type_converters: Dict[str, Callable] = {}
+
+    def __init__(self, entry: dict):
+        self._entry = None
+        self._converted = dict.fromkeys(self._type_converters)
+
+        self.entry = entry
+
+    @property
+    def entry(self):
+        """Get OPTIMADE entry"""
+        return self._entry
+
+    @entry.setter
+    def entry(self, value: dict):
+        """Set OPTIMADE entry
+        If already set, print that this can _only_ be set once.
+        """
+        if self._entry is None:
+            self._entry = self.ENTRY_RESOURCE(**value)
+        else:
+            print("entry can only be set once and is already set.")
+
+    def convert(self, format: str) -> Any:
+        """Convert OPTIMADE entry to desired format"""
+        if format not in self._type_converters:
+            raise AttributeError(
+                f"Non-valid entry type to convert to: {format}. "
+                f"Valid entry types: {tuple(self._type_converters.keys())}"
+            )
+
+        if self._converted[format] is None:
+            self._converted[format] = self._type_converters[format](self.entry)
+
+        return self._converted[format]
+
+    def __getattr__(self, name: str) -> Any:
+        """Get converted entry or attribute from OPTIMADE entry
+        Order:
+        - Try to return converted entry if using `get_<_type_converters value>`.
+        - Try to return OPTIMADE ENTRY_RESOURCE attribute.
+        - Raise AttributeError
+        """
+        # get_<entry_type>
+        if name.startswith("get_"):
+            entry_type = "_".join(name.split("_")[1:])
+            return self.convert(entry_type)
+
+        # Try returning ENTRY_RESOURCE attribute
+        try:
+            res = getattr(self.entry, name)
```

Comment:
I see in the tests that this is intended, so no bother.

Reply:
Yeah, sorry, the `attribute` in the comment line may be confusing here. The intention was just return every property at the toplevel of `EntityResource`, i.e., `id`, `type`, `attributes`, etc.
So if one actually _wants_ a value under `attributes` it would, e.g., go like `entity.attributes.nsites`.

Reply:
Later on we could add more fanciful detection, to first check at the top-level, and later under `attributes` and maybe even under `relationships` or something? So that `entity.nsites` works alongside `entity.id`.

Reply:
> Later on we could add more fanciful detection, to first check at the top-level, and later under `attributes` and maybe even under `relationships` or something? So that `entity.nsites` works alongside `entity.id`.

That's what I thought you were aiming for from the comment (maybe just tweak the wording?), could be useful in the future, I agree.

Reply:
Maybe I could just add it now? Although, I would like to add some proper testing for this as well.. hmm... I'll see what happens.